File: reverse_ll.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	head := ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3}}}
	new_head := reverseList(&head)
	for ; new_head != nil; new_head = new_head.Next {
		fmt.Printf("%d\t", new_head.Val)
	}
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	var temp *ListNode
	for head != nil {
		temp = head.Next
		head.Next = prev
		prev = head
		head = temp
	}
	return prev
}

File: ReverseLL.java
import structs.ListNode;

public class ReverseLL{
    public static void main(String[] args) {
        ListNode ll = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, null))));
        ListNode reversed = new ReverseLL().reverseList(ll);
        while(reversed != null){
            System.out.print(reversed.val + "->");
            reversed = reversed.next;
        }
        System.out.print("\n");
    }

    public ListNode reverseList(ListNode head) {
        ListNode prev = null, temp = null;
        while(head != null){
            temp = head.next;
            head.next = prev;
            prev = head;
            head = temp;
        }
        return prev;
    }
}